pilco/safe_pilco/safe_pilco.py

```diff
@@ -0,0 +1,216 @@
+#         '''
+#         Optimize GP models
+#         '''
+#         self.mgpr.optimize(restarts=restarts)
+#         # Print the resulting model parameters
+#         # ToDo: only do this if verbosity is large enough
+#         lengthscales = {}; variances = {}; noises = {};
+#         i = 0
+#         for model in self.mgpr.models:
+#             lengthscales['GP' + str(i)] = model.kern.lengthscales.value
+#             variances['GP' + str(i)] = np.array([model.kern.variance.value])
+#             noises['GP' + str(i)] = np.array([model.likelihood.variance.value])
+#             i += 1
+#         print('-----Learned models------')
+#         pd.set_option('precision', 3)
+#         print('---Lengthscales---')
+#         print(pd.DataFrame(data=lengthscales))
+#         print('---Variances---')
+#         print(pd.DataFrame(data=variances))
+#         print('---Noises---')
+#         print(pd.DataFrame(data=noises))
+#
+#
+#     def optimize_policy(self, maxiter=50, restarts=1):
+#         '''
+#         Optimize controller's parameter's
+#         '''
+#         start = time.time()
+#         if not self.optimizer:
+#             self.optimizer = gpflow.train.ScipyOptimizer(method="L-BFGS-B")
+#             self.optimizer.minimize(self, maxiter=maxiter)
+#         else:
+#             session = self.optimizer._model.enquire_session(None)
+#             self.optimizer.minimize(self, maxiter=maxiter, session=session)
+#         end = time.time()
+#         print("Controller's optimization: done in %.1f seconds with reward=%.3f." % (end - start, self.compute_reward()))
+#         restarts -= 1
+#
+#         session = self.optimizer._model.enquire_session(None)
+#         best_parameters = self.read_values(session=session)
+#         best_reward = self.compute_reward()
+#         for restart in range(restarts):
+#             print("Before randomisation ", self.compute_reward())
+#             self.controller.randomize()
+#             print("After randomisation", self.compute_reward())
+#             start = time.time()
+#             self.optimizer.minimize(self, maxiter=maxiter, session=session)
+#             end = time.time()
+#             reward = self.compute_reward()
+#             print("Controller's optimization: done in %.1f seconds with reward=%.3f." % (end - start, self.compute_reward()))
+#             if reward > best_reward:
+#                 best_parameters = self.read_values(session=session)
+#                 best_reward = reward
+#
+#         self.assign(best_parameters)
+#         end = time.time()
+#
+#     def training_loss(self):
+#         # This is for tuning controller's parameters
+#         reward = self.predict(self.m_init, self.S_init, self.horizon)[2]
+#         return -reward
+#
+#     def compute_reward(self):
+#         return -self.training_loss()
+#
+#     @property
+#     def maximum_log_likelihood_objective(self):
+#         return -self.training_loss()
+
```

Comment:
I think it's best to delete this and any other debug comments. We can always merge later, if they reach an "uncommented" state.

Reply:
Yup these should definitely be deleted, they are actually an older version, replaced by the SafePILCO class definition underneath